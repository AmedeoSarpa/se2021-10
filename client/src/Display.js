import { useState } from "react";
import { ListGroup, Button, Alert, Row, Col, Card } from "react-bootstrap";
import { Link } from "react-router-dom";

function Display(props) {
  return (
    <>
      {" "}
      <Card body>
        <Card body>
          Now serving...
          <Row className='justify-content-center'>
            {" "}
            <h2> Counter {props.counterCall.counterId} : </h2>
          </Row>
          <Row className='justify-content-center'>
            {" "}
            <h1>{props.counterCall.idTicket} </h1>{" "}
          </Row>
        </Card>

        <Card body className='justify-content-center mt-3'>
          <Row className='justify-content-center'>Waiting list</Row>
          <Row>
            {" "}
            <Col> Service ID </Col> <Col> Description</Col>
            <Col> Queue Lenght</Col>{" "}
          </Row>

          {props.services.map((t) => (
            <Row className='block-example border-top border-dark'>
              {" "}
              <Col> S{t.id} </Col> <Col> {t.description} </Col>{" "}
              <Col>
                {" "}
                {
                  props.queues.filter(
                    (k) => k.idService === t.id && k.processed === 0
                  ).length
                }{" "}
              </Col>{" "}
            </Row>
          ))}
        </Card>
      </Card>
    </>
  );
}

export { Display };
