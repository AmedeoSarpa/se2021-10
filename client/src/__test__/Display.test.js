//UNIT TEST 

//Configuration Enzyme
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
Enzyme.configure({ adapter: new Adapter() });
import { shallow, mount, render } from 'enzyme';
import toJson from 'enzyme-to-json'
//Test every function in Officer.js 
import { Display } from '../Display';
const counterId = 2;
const services = [{
    id: 1,
    timeForPerson: 3,
    description: "send a package",
    name: "service1",
},
{ id: 2, timeForPerson: 5, description: "send a mail", name: "service2" },
{ id: 3, timeForPerson: 4, description: "send money", name: "service3" },
{ id: 4, timeForPerson: 8, description: "receive money", name: "service4" }
]; //servizi

const counterService = [
    { counterId: 1, serviceId: 1 },
    { counterId: 2, serviceId: 2 },
    { counterId: 1, serviceId: 3 },
    { counterId: 1, serviceId: 4 },
    { counterId: 1, serviceId: 2 },
    { counterId: 2, serviceId: 3 },
    { counterId: 2, serviceId: 4 }
]; //servizioOffertoDaiCounter

const queues = [
    { idTicket: 0, idService: 1, processed: 1, counterId: 1 },
    { idTicket: 1, idService: 2, processed: 0, counterId: 2 },
    { idTicket: 2, idService: 2, processed: 0, counterId: 2 },
    { idTicket: 3, idService: 3, processed: 0, counterId: 1 },
    { idTicket: 4, idService: 3, processed: 0, counterId: 1 },
    { idTicket: 5, idService: 4, processed: 0, counterId: 1 }
]; //richiesta di un servizio

const counterCall = {
    counterId: 1,
    idTicket: 0,
};

const isQueueLoading = true;


// Processa la richiesta di un Officer per il prossimo cliente sulla base dei servizi che può offrire
const processRequest = (counterId) => { //AGGIUNGERE API
    //prendere servizi di questo counter , prendere le relative stime e le code
    let servicesId = counterService
        .filter((t) => t.counterId === counterId)
        .map((t) => t.serviceId);
    let maxLength = 0;
    let idServiceToServe = 0;
    let minTime = Infinity;
    let currentTime = 0;
    let currentLength = -1;
    //Per ogni servizio, viene valutata la lunghezza della coda, nel caso risulti essere quella più lunga, viene selezionato il primo numero 
    //disponibile. Il tempo viene utilizzato nel caso due code hanno la stessa lunghezza, viene presa quella con minor tempo.
    for (let id of servicesId) {
        currentTime = services
            .filter((t) => t.id === id)
            .map((t) => t.timeForPerson)[0];
        currentLength = queues.filter(
            (k) => k.idService === id && k.processed === 0
        ).length;
        //Aggiorno i dati o quando ho la cosa più lunga o quando ne ho trovate due uguali(in tal caso deve avere anche tempo più basso rispetto 
        //alla precedente altrimenti non aggiorno e prendo quella di prima)
        if (currentLength > maxLength || (currentLength == maxLength && currentTime < minTime)) {
            minTime = currentTime;
            maxLength = currentLength;
            idServiceToServe = id;
        }
    }

    let arrayServiceWaiting = queues
        .filter((t) => t.idService === idServiceToServe && t.processed == 0)
        .map((t) => t.idTicket);
    let idTicketToServe = Math.min(...arrayServiceWaiting);
    setQueues((oldList) => {
        const list = oldList.map((item) => {
            if (item.idTicket === idTicketToServe) {
                return {
                    idTicket: item.idTicket,
                    processed: 1,
                    idService: item.idService,
                    counterId: item.counterId,
                };
            } else return item;
        });
        return list;
    });
    //console.log(queues);
    setCounterCall({ counterId: counterId, idTicket: idTicketToServe });
};


it("renders display without crashing given the required props", () => {
    const wrapper = shallow(<Display
        counterCall={counterCall}
        services={services}
        queues={queues}
        isQueueLoading={isQueueLoading} />)
    expect(toJson(wrapper)).toMatchSnapshot()
})


it("Officer Page props passed correctly", () => {
    const wrapper = mount(<Display
        counterCall={counterCall}
        services={services}
        queues={queues}
        isQueueLoading={isQueueLoading} />);

    expect(wrapper.props().services).toEqual(services);
    expect(wrapper.props().queues).toEqual(queues);
    expect(wrapper.props().isQueueLoading).toEqual(isQueueLoading);
    expect(wrapper.props().counterCall).toEqual(counterCall);

});


