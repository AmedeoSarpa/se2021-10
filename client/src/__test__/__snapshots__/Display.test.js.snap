// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Test renders display without crashing given the required props 1`] = `
<Fragment>
  <Card
    body={true}
  >
    <Card
      body={true}
    >
      <Row
        className="justify-content-center"
        noGutters={false}
      >
        <h2>
           Now serving 
        </h2>
      </Row>
      <Row
        className="justify-content-center"
        noGutters={false}
      >
        <h2>
           Counter 
          1
           : 
        </h2>
      </Row>
      <Row
        className="justify-content-center"
        noGutters={false}
      >
        <h1>
          0
           
        </h1>
      </Row>
    </Card>
    <Card
      body={true}
      className="justify-content-center mt-3"
    >
      <Row
        className="justify-content-center"
        noGutters={false}
      >
        Waiting list
      </Row>
      <Row
        noGutters={false}
      >
        <Col>
           Service ID 
        </Col>
         
        <Col>
           Description
        </Col>
        <Col>
           Queue Lenght
        </Col>
      </Row>
      <Container
        fluid={false}
      >
        <Spinner
          animation="border"
          variant="primary"
        />
      </Container>
    </Card>
  </Card>
</Fragment>
`;

exports[`renders display without crashing given the required props 1`] = `
<Fragment>
  <Card
    body={true}
  >
    <Card
      body={true}
    >
      <Row
        className="justify-content-center"
        noGutters={false}
      >
        <h2>
           Now serving 
        </h2>
      </Row>
      <Row
        className="justify-content-center"
        noGutters={false}
      >
        <h2>
           Counter 
          1
           : 
        </h2>
      </Row>
      <Row
        className="justify-content-center"
        noGutters={false}
      >
        <h1>
          0
           
        </h1>
      </Row>
    </Card>
    <Card
      body={true}
      className="justify-content-center mt-3"
    >
      <Row
        className="justify-content-center"
        noGutters={false}
      >
        Waiting list
      </Row>
      <Row
        noGutters={false}
      >
        <Col>
           Service ID 
        </Col>
         
        <Col>
           Description
        </Col>
        <Col>
           Queue Lenght
        </Col>
      </Row>
      <Container
        fluid={false}
      >
        <Spinner
          animation="border"
          variant="primary"
        />
      </Container>
    </Card>
  </Card>
</Fragment>
`;

exports[`renders without crashing given the required props 1`] = `
<Fragment>
  <Card
    body={true}
  >
    <Card
      body={true}
    >
      <Row
        className="justify-content-center"
        noGutters={false}
      >
        <h1>
           Nobody to call 
        </h1>
      </Row>
    </Card>
    <Card
      body={true}
      className="justify-content-center mt-3"
    >
      <Row
        className="justify-content-center"
        noGutters={false}
      >
        Waiting list
      </Row>
      <Row
        noGutters={false}
      >
        <Col>
           Service ID 
        </Col>
         
        <Col>
           Description
        </Col>
        <Col>
           Queue Lenght
        </Col>
      </Row>
      <Row
        className="block-example border-top border-dark"
        noGutters={false}
      >
        <Col>
           S
          1
           
        </Col>
         
        <Col>
           
          send a package
           
        </Col>
        <Col>
          0
        </Col>
      </Row>
      <Row
        className="block-example border-top border-dark"
        noGutters={false}
      >
        <Col>
           S
          2
           
        </Col>
         
        <Col>
           
          send a mail
           
        </Col>
        <Col>
          2
        </Col>
      </Row>
      <Row
        className="block-example border-top border-dark"
        noGutters={false}
      >
        <Col>
           S
          3
           
        </Col>
         
        <Col>
           
          send money
           
        </Col>
        <Col>
          2
        </Col>
      </Row>
      <Row
        className="block-example border-top border-dark"
        noGutters={false}
      >
        <Col>
           S
          4
           
        </Col>
         
        <Col>
           
          receive money
           
        </Col>
        <Col>
          1
        </Col>
      </Row>
    </Card>
  </Card>
</Fragment>
`;
